### PR TITLE
Added retry checker to http retrying client

### PR DESCRIPTION
**Pull request by kaikash**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

### CODE REVIEW

File: yt/yt/core/http/retrying_client.cpp [hunk 1]
```diff
@@ -18,14 +19,46 @@ static const auto& Logger = HttpLogger;
...
+    }
+
+    virtual TError CheckError(const IResponsePtr& response) override = 0;
+    virtual NYTree::INodePtr GetFormattedResponse() const override = 0;
+
+protected:
+    static bool DefaultRetryChecker(const TError& error)
+    {
+        // TODO: default retry checker
+        Y_UNUSED(error);
```
Comment (Kontakter) [inline yt/yt/core/http/retrying_client.cpp:42]: We prefer to hide name in the method declaration, i.e:
```
static bool DefaultRetryChecker(const TError& /*error*/)
```

File: yt/yt/core/http/retrying_client.cpp [hunk 2]
```diff
@@ -2,6 +2,7 @@
 
 #include "retrying_client.h"
 #include "private.h"
+#include "util/system/compiler.h"
```
[the remark this answers is not in the document]
Comment (kaikash) [reply yt/yt/core/http/retrying_client.cpp:6]: my IDE accidentally added this, these includes are not needed

File: yt/yt/core/http/retrying_client.cpp (hunk 1 above)
Comment (kaikash) [reply yt/yt/core/http/retrying_client.cpp]: This is actually not method declaration. There should be a default retry checking logic. Added in new commit.

File: yt/yt/core/http/retrying_client.cpp [hunk 3]
```diff
@@ -166,49 +198,51 @@ class TRetryingClient
...
-            auto now = TInstant::Now();
-            if (now > deadline) {
+            if (!shouldRetry(checkError)) {
                 break;
             }
+
+            auto now = TInstant::Now();
             TDelayedExecutor::WaitForDuration(std::min(Config_->BackoffTimeout, deadline - now));
-        }
+        } while(true);
```
Comment (Kontakter) [inline yt/yt/core/http/retrying_client.cpp:245]: Cosmetic thing: we usually write infinite loops as
```
while (true) {
}
```

File: yt/yt/core/http/retrying_client.cpp [hunk 4]
```diff
@@ -18,14 +18,44 @@ static const auto& Logger = HttpLogger;
...
+        return RetryChecker_(error);
+    }
+
+    virtual TError CheckError(const IResponsePtr& response) override = 0;
+    virtual NYTree::INodePtr GetFormattedResponse() const override = 0;
+
+protected:
+    static bool DefaultRetryChecker(const TError& error)
+    {
+        return !error.FindMatching(NRpc::EErrorCode::InvalidCredentials).has_value();
```
Comment (Kontakter) [inline yt/yt/core/http/retrying_client.cpp:40]: It looks a little bit weird in the base class, I suppose that here we should just `return true`

But in oauth_service.cpp we can specify checker that make retries for all errors expect InvalidCredentials

**Comment by Kontakter** [comment]
Shipped!

**Pull request closed without merging**